This is notebook file is the Solution for Exercise 2 from the [exercise series 3](https://github.com/righetti/optlearningcontrol/tree/master/Spring2020/series3) for the course Optimal and Learning Control (ME-GY 7973), as solved by Tarun Thathvik Paladugu.

In [1]:
import numpy as np

In [2]:
def find_policy(J, possible_moves):
    A = np.zeros([J.size,J.size])
    for i in range(J.size):
        low = np.inf
        for j in range(possible_moves[i].size):
            if possible_moves[i][j] == 1:
                if low > J[j]:
                    low = J[j]
                    A[i] = np.zeros([A[i].size])
                    A[i,j] = 1
    
    return A

In [3]:
#defining all the unchanged variables for question 1 and 2
g = np.array(([0,0,1,0,-1,0,1,0,10,0,10,1,0,1,0,0,10,-1]))
alpha = 0.99
possible_moves = np.zeros([18,18])

possible_moves[0,0] = 1
possible_moves[0,1] = 1
possible_moves[0,5] = 1
possible_moves[1,0] = 1
possible_moves[1,1] = 1
possible_moves[1,2] = 1
possible_moves[2,1] = 1
possible_moves[2,2] = 1
possible_moves[2,3] = 1
possible_moves[2,6] = 1
possible_moves[3,2] = 1
possible_moves[3,3] = 1
possible_moves[3,4] = 1
possible_moves[3,7] = 1
possible_moves[4,3] = 1
possible_moves[4,4] = 1
possible_moves[4,8] = 1
possible_moves[5,0] = 1
possible_moves[5,5] = 1
possible_moves[5,9] = 1
possible_moves[6,2] = 1
possible_moves[6,6] = 1
possible_moves[6,7] = 1
possible_moves[7,3] = 1
possible_moves[7,6] = 1
possible_moves[7,7] = 1
possible_moves[7,8] = 1
possible_moves[7,11] = 1
possible_moves[8,4] = 1
possible_moves[8,7] = 1
possible_moves[8,8] = 1
possible_moves[8,12] = 1
possible_moves[9,5] = 1
possible_moves[9,9] = 1
possible_moves[9,10] = 1
possible_moves[9,13] = 1
possible_moves[10,9] = 1
possible_moves[10,10] = 1
possible_moves[10,14] = 1
possible_moves[11,7] = 1
possible_moves[11,11] = 1
possible_moves[11,12] = 1
possible_moves[11,16] = 1
possible_moves[12,8] = 1
possible_moves[12,11] = 1
possible_moves[12,12] = 1
possible_moves[12,17] = 1
possible_moves[13,9] = 1
possible_moves[13,13] = 1
possible_moves[13,14] = 1
possible_moves[14,10] = 1
possible_moves[14,13] = 1
possible_moves[14,14] = 1
possible_moves[14,15] = 1
possible_moves[15,14] = 1
possible_moves[15,15] = 1
possible_moves[15,16] = 1
possible_moves[16,11] = 1
possible_moves[16,15] = 1
possible_moves[16,16] = 1
possible_moves[16,17] = 1
possible_moves[17,12] = 1
possible_moves[17,16] = 1
possible_moves[17,17] = 1

## Value Iteration

### Part a)

In [4]:
#Value Iteration

iteration = 1
change = 10**(-6)

#defining the initial J
J = np.full(g.shape, 0)


while(1):
    #while
    next_step = np.full(J.shape, np.inf)
    for i in range(possible_moves[0].size):
        for j in range(possible_moves[i].size):
            if possible_moves[i][j] == 1:
                if next_step[i] > J[j]:
                    next_step[i] = J[j]
    J_old = J
    J = g + alpha*next_step
    if all(abs(J - J_old) < change):
        break
    iteration += 1
    
print("The number of Iterations until convergance = ", iteration)
print("The final cost = ")
J_a = np.full(J.size, J)
print(J_a)

The number of Iterations until convergance =  1376
The final cost = 
[-95.07940237 -96.03980137 -97.00990137 -98.99990137 -99.99990137
 -94.12860736 -96.02980137 -98.00990137 -88.99990137 -93.1873203
 -82.25544611 -97.00990137 -98.99990137 -91.25544611 -90.34289066
 -89.43946077 -88.99990137 -99.99990137]


## Policy Iteration

### Part b)

In [5]:
#Initiating an arbitrary A that does not move
A_old = np.eye(18)

#Initiating variable to keep count of Iterations
iteration = 1

#change = 0.000001  #Used when checking for the convergance of J
while(1):
    J = np.linalg.solve(np.eye(18) - alpha * A_old, g)
    A = find_policy(J,possible_moves)     #Finding the new A from the new J
    #if all(abs(J - J_old) <= change):
    if (A == A_old).all():
        break
    #J_old = J
    A_old = np.full(A.shape, A)
    iteration += 1

print("The number of Iterations until convergance = ", iteration)
print("The final cost = ")
J_b = np.full(J.size, J)
print(J_b)

print()
print("The final Policy after Policy Iteration: ")
print(A)

The number of Iterations until convergance =  8
The final cost = 
[ -95.079501    -96.0399      -97.01        -99.         -100.
  -94.12870599  -96.0299      -98.01        -89.          -93.18741893
  -82.25554474  -97.01        -99.          -91.25554474  -90.34298929
  -89.4395594   -89.         -100.        ]

The final Policy after Policy Iteration: 
[[0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0

### Part c)

Solving Linear Algebra and the large size of the A matrix might make Policy Iteration slightly more computationally intensive in each step, but the total number of iterations taken to converge to a desired Policy is much greater for Value Iteration which makes the overall time taken much higher for the value iteration (when run, it was found that the time taken to run the Policy Iteration is .63s while the time taken to run the value Iteration is .32s). The policies in Policy Iteration converges at the 8th iteration, while it takes 1376 iterations for the cost function to converge through the value iteration converges. From it is clear that 1376 iterations of value iteration took only twice as long as the 8 iterations of the Policy Iteration algorithm. Therefore, the Policy Iteration Algorithm is computationally intensive but gives faster results, while the Value Iteration is relatively very light for computation, yet slow to converge. It is a faor comparison since the initail conditions for both the value iteration and policy iteration are the same (Zero Value inital condition is the same as the policy that does not move). 

In other words, the quick convergence of the Policy Iteration more than compensates for the higher complexity due to linear algebra. And due to the slow convergence of the Value Iteration, it takes much longer to compute although the complexity of the algorithm is very low.

<p> However, despite the large differences in the number of iterations and time consumption, the J from Value iteration and J$_{\mu}$ from Policy iteration are almost equal. And therefore the policies are both equal.

In [6]:
print("J from Value Iteration: ")
print(J_a)
print("J from Policy Iteration: ")
print(J_b)

print()
print("All the values of both the policies are equall : ", (find_policy(J_a, possible_moves)==find_policy(J_b, possible_moves)).all())

J from Value Iteration: 
[-95.07940237 -96.03980137 -97.00990137 -98.99990137 -99.99990137
 -94.12860736 -96.02980137 -98.00990137 -88.99990137 -93.1873203
 -82.25544611 -97.00990137 -98.99990137 -91.25544611 -90.34289066
 -89.43946077 -88.99990137 -99.99990137]
J from Policy Iteration: 
[ -95.079501    -96.0399      -97.01        -99.         -100.
  -94.12870599  -96.0299      -98.01        -89.          -93.18741893
  -82.25554474  -97.01        -99.          -91.25554474  -90.34298929
  -89.4395594   -89.         -100.        ]

All the values of both the policies are equall :  True
